src/base/multithreading.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2020 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#ifndef WL_BASE_MULTITHREADING_H
+#define WL_BASE_MULTITHREADING_H
+
+#include <memory>
+#include <mutex>
+#include <thread>
+
+#include "notifications/note_ids.h"
+#include "notifications/notifications.h"
+
+void set_initializer_thread();
+bool is_initializer_thread();
+
+/*
+ * Informs the drawing thread to run the given function ASAP.
+ * This must not be instantiated directly – use `instantiate()` instead,
+ * which will also take care of publishing the note.
+ */
+struct NoteThreadSafeFunction {
+	CAN_BE_SENT_AS_NOTE(NoteId::ThreadSafeFunction)
+	static void instantiate(const std::function<void()>&, bool wait_until_completion);
+
+	const std::function<void()> run;
+
+	const uint32_t id;
+
+private:
+	NoteThreadSafeFunction(const std::function<void()>& f) : run(f), id(next_id_++) {
+	}
+
+	static uint32_t next_id_;
+};
+
+// Used only by Panel code
+struct NoteThreadSafeFunctionHandled {
+	const uint32_t id;
+	explicit NoteThreadSafeFunctionHandled(uint32_t i) : id(i) {
+	}
+	CAN_BE_SENT_AS_NOTE(NoteId::ThreadSafeFunctionHandled)
+};
+
+/* Ensures that critical pieces of code are executed by only one thread at a time.
+ * More precisely: If n pieces of code C1,…,Cn are structured like this:
+ *    {
+ *        MutexLock m(...);
+ *        do_something();
+ *    }
+ * Then there will never, at any given time, two threads T1 and T2
+ * be running any pieces Cx and Cy (1 <= x,y <= n) concurrently.
+ * Each thread may, of course, call Cx from within Cy without problems.
+ *
+ * Use MutexLocks sparingly. They are there to safeguard bottlenecks in the
+ * code where concurrency must not be allowed, at the cost of performance.
+ */
+class MutexLock {
+public:
+	// Which mutex to lock. Each entry corresponds to a different mutex.
+	enum class ID { kLogicFrame, kObjects, kCommands, kMessages, kIBaseVisualizations };
```

Comment:
Is there a 1<->1 relation to instabces of std::thread ? 
If so you should make this visible in some way

Reply:
There's no relation to threads. The point is that there are some code snippets that must not be run concurrently (by another thread) with some other code snippets, so those pairs of code pieces are surrounded with a lock for the same mutex to prevent concurrency of problematic pairs of code.